iocore/net/SSLNetVConnection.cc

```diff
@@ -603,11 +603,20 @@ SSLNetVConnection::net_read_io(NetHandler *nh, EThread *lthread)
   if (!getSSLHandShakeComplete()) {
     int err = 0;
 
+    // May get into logic that will clean up the current VC
+    // Increment the recursion to delay do_io_close cleaup.
+    this->recursion++;
+
     if (get_context() == NET_VCONNECTION_OUT) {
       ret = sslStartHandShake(SSL_EVENT_CLIENT, err);
     } else {
       ret = sslStartHandShake(SSL_EVENT_SERVER, err);
     }
+    if (ret == EVENT_RESTART) {
+      // VC migrated into a new object
+      // Just give up and go home. Events should trigger on the new vc
+      return;
```

Comment:
Don't we need to decrement the recursion counter here?

If no, it's difficult to understand. Want a comment why we don't need it.
If yes, it's fragile. I don't like to use `goto` but we may need it use it.

Reply:
I tried the recursion counts then changed the return value.  I think that just sending back a different return value will suffice.  I'll back out the recursion could and verify that everything still works.  And update the PR

Reply:
And you are exactly right.  Should be dropping the recursion count here too.  Should use the scoped pattern to deal with the decrement if we really need to manipulate the recursion counter.